src/Nethermind/Nethermind.Merge.Plugin/GC/GCKeeper.cs

```diff
@@ -0,0 +1,119 @@
+                {
+                    failCause = FailCause.GCFailedToStartNoGCRegion;
+                }
+            }
+            catch (ArgumentOutOfRangeException)
+            {
+                failCause = FailCause.TotalSizeExceededTheEphemeralSegmentSize;
+            }
+            catch (InvalidOperationException)
+            {
+                failCause = FailCause.AlreadyInNoGCRegion;
+            }
+
+            return new NoGCRegion(this, failCause, size, _logger);
+        }
+
+        return new NoGCRegion(this, FailCause.StrategyDisallowed, size, _logger);
+    }
+
+    private enum FailCause
+    {
+        None,
+        StrategyDisallowed,
+        GCFailedToStartNoGCRegion,
+        TotalSizeExceededTheEphemeralSegmentSize,
+        AlreadyInNoGCRegion
+    }
+
+    private class NoGCRegion : IDisposable
+    {
+        private readonly GCKeeper _gcKeeper;
+        private readonly FailCause _failCause;
+        private readonly long? _size;
+        private readonly ILogger _logger;
+
+        internal NoGCRegion(GCKeeper gcKeeper, FailCause failCause, long? size, ILogger logger)
+        {
+            _gcKeeper = gcKeeper;
+            _failCause = failCause;
+            _size = size;
+            _logger = logger;
+        }
+
+        public void Dispose()
+        {
+            if (_failCause == FailCause.None)
+            {
+                if (GCSettings.LatencyMode == GCLatencyMode.NoGCRegion)
+                {
+                    try
+                    {
+                        System.GC.EndNoGCRegion();
+                    }
+                    catch (InvalidOperationException)
+                    {
+                        if (_logger.IsDebug) _logger.Debug($"Failed to keep in NoGCRegion with Exception with {_size} bytes");
+                    }
+                }
+                else if (_logger.IsDebug) _logger.Debug($"Failed to keep in NoGCRegion with {_size} bytes");
+            }
+            else if (_logger.IsDebug) _logger.Debug($"Failed to start NoGCRegion with {_size} bytes with cause {_failCause.FastToString()}");
+
+            _gcKeeper.ScheduleGC();
+        }
+    }
+
+    private void ScheduleGC()
+    {
+        if (_gcScheduleTask.IsCompleted) _gcScheduleTask = ScheduleGCInternal();
+    }
+
+    private async Task ScheduleGCInternal()
+    {
+        (int generation, bool compacting) = _igcStrategy.GetForcedGCParams();
+        if (generation >= 0)
+        {
+            // This should give time to finalize response in Engine API
+            // Normally we should get block every 12s (5s on some chains)
+            // Lets say we process block in 2s, then delay 1s, then invoke GC
+            await Task.Delay(1000);
+            if (_logger.IsDebug) _logger.Debug($"Forcing GC collection of gen {generation}, compacting {compacting}");
+            System.GC.Collect(generation, GCCollectionMode.Default, blocking: false, compacting: compacting);
```

Comment:
```suggestion
            System.GC.Collect(generation, GCCollectionMode.Default, blocking: false, compacting: false);
            System.GC.WaitForPendingFinalizers();
            System.GC.Collect(generation, GCCollectionMode.Default, blocking: false, compacting: compacting);
            
```

Reply:
We don't have any explicit finalizers in our code, so it would have been something from libraries or runtime

Reply:
`EthashCache` has one, not sure what used for 🤷‍♂️ 

Anything with finalizer will be released on first pass, and then turn to garbage on second, so second sweep is to remove them entirely. If you are manually inducing can be worthwhile

Reply:
IMHO it's overkill here to force two GCs because of finalizers.